crates/red_knot_python_semantic/resources/mdtest/attributes.md

```diff
@@ -112,45 +130,66 @@ We also recognize pure instance variables if they are defined in a method that i
+        self.declared_and_bound: bool = True
+
+c_instance = C(1)
+
+# TODO: Should be `Unknown | Literal[1, "a"]`
+reveal_type(c_instance.inferred_from_value)  # revealed: @Todo(implicit instance attribute)
+
+# TODO: Should be `Unknown | Literal[1, "a"]`
+reveal_type(c_instance.inferred_from_other_attribute)  # revealed: @Todo(implicit instance attribute)
 
-# Not that we would use this in static analysis, but for a more realistic example, let's actually
-# call the method, so that the attribute is bound if this example is actually run.
-c_instance.set_instance_variable()
+# TODO: Should be `int | None`
+reveal_type(c_instance.inferred_from_param)  # revealed: @Todo(implicit instance attribute)
 
-# TODO: should be `Literal["value set in method"]` or `Unknown | Literal[…]` (see above).
-reveal_type(c_instance.pure_instance_variable)  # revealed: @Todo(implicit instance attribute)
+# TODO: Should be `bytes`
+reveal_type(c_instance.declared_only)  # revealed: @Todo(implicit instance attribute)
+
+# TODO: Should be `bool`
+reveal_type(c_instance.declared_and_bound)  # revealed: @Todo(implicit instance attribute)
 
 # TODO: We already show an error here, but the message might be improved?
 # error: [unresolved-attribute]
-reveal_type(C.pure_instance_variable)  # revealed: Unknown
+reveal_type(C.inferred_from_value)  # revealed: Unknown
 
 # TODO: this should be an error
-C.pure_instance_variable = "overwritten on class"
+C.inferred_from_value = "overwritten on class"
 ```
 
-#### Variable declared in class body and not bound anywhere
-
-If a variable is declared in the class body but not bound anywhere, we still consider it a pure
-instance variable and allow access to it via instances.
+#### Methods that does not use `self` as a first parameter
 
 ```py
 class C:
-    pure_instance_variable: str
+    # This might trigger a stylistic lint like `invalid-first-argument-name-for-method`, but
+    # it should be supported in general:
+    def __init__(this) -> None:
+        this.declared_and_bound: str | None = "a"
 
-c_instance = C()
+# TODO: should be `str | None`
+reveal_type(C().declared_and_bound)  # revealed: @Todo(implicit instance attribute)
+```
 
-reveal_type(c_instance.pure_instance_variable)  # revealed: str
+#### Aliased `self` parameter
 
-# TODO: mypy and pyright do not show an error here, but we plan to emit a diagnostic.
-# The type could be changed to 'Unknown' if we decide to emit an error?
-reveal_type(C.pure_instance_variable)  # revealed: str
+```py
+class C:
+    def __init__(self) -> None:
+        this = self
+        this.declared_and_bound: str | None = "a"
 
-# TODO: mypy and pyright do not show an error here, but we plan to emit one.
-C.pure_instance_variable = "overwritten on class"
+# TODO: This would ideally be `str | None`, but mypy/pyright don't support this either,
+# so `Unknown` + a diagnostic is also fine.
+reveal_type(C().declared_and_bound)  # revealed: @Todo(implicit instance attribute)
```

Comment:
ooh, interesting. Yeah, I can see how this could be hard. Simply checking that the `this` variable in `__init__` has the same type as `self` would not be sufficient; you'd need to know that it refers to exactly the same object as `self`.

Reply:
Interestingly, pyright does actually seem to have [a special treatment of `self` / `cls`](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=inferred-type-of-self-and-cls-parameters):

> When a type annotation for a method’s `self` or `cls` parameter is omitted, pyright will infer its type based on the class that contains the method. The inferred type is internally represented as a type variable that is bound to the class.
>
> The type of `self` is represented as `Self@ClassName` where `ClassName` is the class that contains the method. Likewise, the `cls` parameter in a class method will have the type `Type[Self@ClassName]`.

```py
class C:
    def __init__(self) -> None:
        reveal_type(self)  # pyright:  Self@C
```

Reply:
If by "special treatment" you mean "special treatment based on the name of the parameter," I don't see evidence for that in the linked docs or in the playground? Pyright does implement a lint rule telling you to change the name to `self`/`cls`, but using a different name doesn't seem to otherwise impact its type inference. The phrase "a method's self or cls parameter" in those docs does not mean "a parameter named `self` or `cls`", it means "the first parameter of a method or classmethod."

If by "special treatment" you mean assigning a type to an un-annotated first parameter that is a typevar bound to the type of the containing class, we will also need to do that.

Reply:
> Simply checking that the `this` variable in `__init__` has the same type as `self` would not be sufficient; you'd need to know that it refers to exactly the same object as `self`.

Once we implement correct `Self` typing, then checking that it has the same type as `self` will actually be sufficient, because the `Self` type can't apply to anything other than `self`.

Reply:
> Once we implement correct `Self` typing, then checking that it has the same type as `self` will actually be sufficient, because the `Self` type can't apply to anything other than `self`.

Is that true?

```py
from typing import Self

class Foo:
    def __init__(self, other: Self | None = None):
        if other is not None:
            other.bar = 42
```

Reply:
Hmm, good point. I think it would also be pretty reasonable to consider `other.bar = 42` the same way we would consider `self.bar = 42`; both are cases of a method of the very same class assigning to an attribute on its self type. The only real distinction would be if we were trying to validate full initialization of `self`, but I don't think we're aiming to do that? But it would also be OK not to, whatever is easier I think, since I doubt it happens often.

Reply:
Ah, yeah, that makes sense!

Reply:
> If by "special treatment" you mean assigning a type to an un-annotated first parameter that is a typevar bound to the type of the containing class, we will also need to do that.

Sorry, I should have been more clear. This is what I meant, yes. It's not yet clear to me why pyright sometimes infers `C`, [sometimes `*C`](https://pyright-play.net/?strict=true&code=MYGwhgzhAEDCBcAoaLoBMCmAzaB9XAlgHYEAu%2BAFBBiFgJTQC0AfNAHID2RGSqf01WgDoAHtAC8AmliEBPIA) (for "partially unknown" types?), and sometimes `Self@C`.

Reply:
Ah, I see. The type of `self` should generally be `Self@C`, which is a type variable with upper bound `C`; this is different from `C` in that it accounts for the possibility of a subclass, which has implications for methods that  `return self` (when called on a subclass, they will return an instance of the subclass, not `C`), and for valid-Liskov-override checking of method signatures (normally parameter types are contravariant, so a subclass method parameter could not accept a subtype of the corresponding superclass method parameter, but this is exactly what needs to happen with `self`). Not sure in what cases pyright will flatten that to just `C`, or what `C*` represents.

Reply:
some of the inconsistency in pyright's behaviour here may be because of what PEP484 specified before we added a `Self` type to the typing system: https://peps.python.org/pep-0484/#annotating-instance-and-class-methods

Pyright and other type checkers had to change various behaviours when `Self` was added